Add config that allows Elasticsearch mapper parsing errors to be ignored

zzbennett opened the pull request.

ConfluentJenkins commented:
Can one of the admins verify this patch?

nabeelz6 commented:
Any chance this could be merged? This is a nice option to have for what we're doing. @zzbennett - not sure if you need to fix the merge conflict.

zzbennett commented:
@nabeelz6 sure, I can fix the merge conflict. I've since made a couple more additions to this patch so I'll add those here as well. We've been running with this patch in prod for the last three months and it's made the difference between not being able to use the ES connector and being able to run it just fine and never having to think about it.

nabeelz6 commented:
@zzbennett - Great! So the errors will be ignored with this patch but elasticsearch still won't process that document / batch of documents? Or would elasticsearch still process that document but ignore the field(s) that caused the issue?

zzbennett commented:
@nabeelz6 the way it's written now, it'll just log the document and drop it, which is pretty extreme but for our use case, it doesn't matter. Other documents in the same batch will still be indexed. There are a lot of options for how to handle the errors so that the document can still be indexed. Is there something that would work particularly well for your use case?

nabeelz6 commented:
@zzbennett Ah okay. Ideally, we'd like the document with `mapper_parsing_exceptions` to still make it into ES index.. the property that caused the issue can be excluded from index if need be. 

We tried to configure the ES index for `ignore-malformed` but it didn't seem to work when a property that was an object in one document later appears in another document as a string.

rhauch commented on src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java:
```diff
@@ -80,7 +81,10 @@ protected static ConfigDef baseConfigDef() {
+                  group, ++order, Width.SHORT, "Retry Backoff (ms)")
+          .define(IGNORE_MAPPING_ERRORS_CONFIG, Type.BOOLEAN, false, Importance.LOW,
+          "Whether to ignore mapper_parsing_exceptions thrown by Elasticsearch upon indexing malformed documents.",
+                  group, ++order, Width.SHORT, "Ignore Mapping Errors mode");
```
Do we really want the documentation to include the `mapper_parsing_exceptions` term, or would it be better to remove the underscores?

Also, every time we update or change any of the configuration files, we'd like the PR to also update the `docs/configuration_options.rst` contents. The easiest way to do that is to run the [`main(...)` method in the `ElasticsearchSinkConnectorConfig`](https://github.com/stitchfix/kafka-connect-elasticsearch/blob/master/src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java#L295-L297) class, and they carefully copy/paste the contents accordingly.

I do this through my IDE or on the command line with:

    mvn exec:java -Dexec.mainClass="io.confluent.connect.elasticsearch/ElasticsearchSinkConnectorConfig" -Dexec.classpathScope="test"

rhauch replied:
Actually, after thinking about this a bit more, using boolean here is simple but it offers no room for handling other behaviors. Have you thought about using an enumeration to define the different behaviors, similar to the `BehaviorOnNullValues` enum used in #165? The advantage is that we could define other behaviors besides "ignore", such as "fail" and "write" (as suggested by @nabeelz6).

zzbennett replied:
Thanks for the review @rhauch. I like the idea of being able to configure different options for what to do when a conflict happens. I'll take a look at #165 for inspiration and update the PR. Will also update the `config_options.rst`

nabeelz6 replied:
Thanks @zzbennett!

zzbennett replied:
FYI I went on vacation for a week and am recovering from the flu now, but I am still planning on implementing the suggestion! Shooting for middle of next week.

zzbennett commented:
Rebased to fix merge conflicts and make it current with master. Am working on implementing  @rhauch suggestions now.

zzbennett commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -391,6 +404,38 @@ private BulkResponse execute() throws Exception {
+      // the behavior is configurable.
+      switch (behaviorOnMalformedDoc) {
+        case WARN:
+          log.warn("Encountered mapper_parsing_exception when executing batch {} of {}"
```
will fix this log statement

zzbennett commented:
Okay, I've tidied it up and implemented @rhauch suggestions. I'm open to more feedback, especially ideas for how to somehow re-index/preserve the documents that had issues, but at least there's an easy path forward now to later add that functionality.

I'm not sure why the build is failing, I can't get the jenkins to load.

rhauch commented:
@rayokota would you mind taking a look?

kkonstantine commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -419,4 +472,51 @@ private static ConnectException toConnectException(Throwable t) {
+
+    public static final BehaviorOnMalformedDoc DEFAULT = FAIL;
+
+    // Want values for "behavior.on.null.values" property to be case-insensitive
```
remainder from the other enum

kkonstantine commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -372,9 +378,17 @@ private BulkResponse execute() throws Exception {
                       batchId, batch.size(), attempts, maxAttempts);
             }
             return bulkRsp;
+          } else {
```
`else if(...)`
then
`else`
No need for the extra level of nesting

kkonstantine commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -391,6 +405,45 @@ private BulkResponse execute() throws Exception {
+              batchId, batch.size(), bulkRsp.getErrorInfo());
+          return;
+        case IGNORE:
+          if (log.isDebugEnabled()) {
```
You probably can skip this check. As elsewhere in the code, the log level is not checked when `batch.size()` is called, because the collection is an array list. Since calling `getErrorInfo` is also negligible we can skip the extra code here.

kkonstantine commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -391,6 +405,45 @@ private BulkResponse execute() throws Exception {
+        case IGNORE:
+          if (log.isDebugEnabled()) {
+            log.debug("Encountered an illegal document error when executing batch {} of {}"
+                    + " records. Ignoring. Error was {}",
```
Ideally it would match the format of the message above. `Error was {}. Ignoring and will not index record.`

kkonstantine commented on src/test/java/io/confluent/connect/elasticsearch/bulk/BulkProcessorTest.java:
```diff
@@ -305,4 +319,84 @@ public void unretriableErrors() throws InterruptedException {
+    client.expect(Arrays.asList(42, 43), BulkResponse.failure(false, errorInfo));
+
+    final BulkProcessor<Integer, ?> bulkProcessor = new BulkProcessor<>(
+        new SystemTime(),
```
optional: I know it's same as everywhere in this test, but should we use `Time.SYSTEM` instead? Feel free to change everywhere, only in new tests, or not at all. Kind of like [ignore, warn, fail]-style set of options here too :)

kkonstantine commented on src/test/java/io/confluent/connect/elasticsearch/bulk/BulkProcessorTest.java:
```diff
@@ -305,4 +319,84 @@ public void unretriableErrors() throws InterruptedException {
+    bulkProcessor.add(42, 1);
+    bulkProcessor.add(43, 1);
+
+
```
nit: extra line

kkonstantine commented on src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:
```diff
@@ -391,6 +405,45 @@ private BulkResponse execute() throws Exception {
+      // if the elasticsearch request failed because of a malformed document,
+      // the behavior is configurable.
+      switch (behaviorOnMalformedDoc) {
+        case WARN:
```
very nitpicky but I wouldn't mind matching the order of declaration in the enum (since order doesn't really matter here). `IGNORE, WARN, FAIL`

kkonstantine commented:
Also, the new unit tests failed.

zzbennett commented:
Thanks @kkonstantine for the additional suggestions. Love the attention to detail you Kafka folks have 😄 I will implement the suggestions and update the PR.

zzbennett commented:
I've updated with the review feedback and also modified the test slightly to ensure we test all three options, even though IGNORE and WARN are logically equivalent.

I ran `mvn install` locally with the latest snapshots and all the unit tests pass, so I'm not sure what the issue with the testing is. @kkonstantine do you have the specific details of the test failure?

rayokota commented:
Sorry for the late response.  I took a look at the changes and LGTM.

rhauch commented:
@zzbennett The most recent CI failure is due to a test failure:

```
java.lang.AssertionError
	at io.confluent.connect.elasticsearch.bulk.BulkProcessorTest.failOnMalformedDoc(BulkProcessorTest.java:361)
Standard Output
[2018-02-22 01:00:59,392] ERROR Encountered an illegal document error when executing batch 2 of 2 records. Error was  [{"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"illegal_argument_exception","reason":"object
 field starting or ending with a [.] makes object resolution ambiguous: [avjpz{{.}}wjzse{{..}}gal9d]"}}] (to ignore future records like this change the configuration property '%s' from '%s' to '%s'). (io.confluent.connect.elasticsearch.bulk.BulkProcessor:422)
[2018-02-22 01:00:59,401] ERROR Failed to execute batch 2 of 2 records after total of 1 attempt(s) (io.confluent.connect.elasticsearch.bulk.BulkProcessor:399)
org.apache.kafka.connect.errors.ConnectException: Bulk request failed:  [{"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"illegal_argument_exception","reason":"object
 field starting or ending with a [.] makes object resolution ambiguous: [avjpz{{.}}wjzse{{..}}gal9d]"}}]
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.handleMalformedDoc(BulkProcessor.java:429)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.execute(BulkProcessor.java:383)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:344)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:330)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
```

kkonstantine commented:
Thanks @zzbennett for addressing the final comments (nitpicks included)! 

Tests were failing on the assertion of the exception message with: 
`Run 1: BulkProcessorTest.ignoreOrWarnOnMalformedDoc:409 Flush timeout expired with unflushed records: ` 

The flush timeout was too short and `ConnectException` would be thrown, which is the expected exception, but would have different reason than the expected one. Increasing the flush timeout passed to `flush` makes the tests non-flaky. 

I hope you don't mind me pushing this small change and unblocking this PR so we can merge it. I'll also rebase on top of `4.1.x`

kkonstantine commented:
Rebase and build successful. Merging to `4.1.x`. Thanks @zzbennett !
